File: packages/senri_ffi_js/src/callback.ts
/*
 * Copyright (c) 2026 TT23XR Studio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Pointer } from './pointer';
import { FFIError } from './errors';
import { getGlobalAdapter } from './globals';
import { getResourceRegistry } from './resource-registry';
import { normalizeType } from './types/normalized';

let _finalizationRegistry: FinalizationRegistry<object> | null = null;

export function callback(retType: any, argTypes: any[], jsFn: Function, options?: any): Pointer {
  if (typeof jsFn !== 'function') throw new FFIError('callback requires a function');

  const adapter = getGlobalAdapter();
  const normalizedRet = normalizeType(retType);
  const normalizedArgs = argTypes.map(t => normalizeType(t));

  const desc = adapter.registerCallback(jsFn, normalizedRet, normalizedArgs);
  (desc as any)._unregister = () => {
    getResourceRegistry().unregisterCallback(desc);
    adapter.unregisterCallback(desc);
  };
  getResourceRegistry().registerCallback(desc);

  const pointer = new Pointer(desc);

  if (typeof FinalizationRegistry !== 'undefined') {
    if (!_finalizationRegistry) {
      _finalizationRegistry = new FinalizationRegistry((held: any) => {
        if (held && typeof held._unregister === 'function') {
          try { held._unregister(); } catch {}
        }
      });
    }
    _finalizationRegistry.register(pointer, desc);
  }

  return pointer;
}

File: packages/senri_ffi_py/src/senri_ffi/callback.py
from __future__ import annotations

import weakref
from typing import Any, Callable

from senri_ffi.errors import FFIError
from senri_ffi.globals import get_global_adapter
from senri_ffi.pointer import Pointer
from senri_ffi.registry import get_resource_registry
from senri_ffi.types.normalized import normalize_type

_finalization_registry: weakref.WeakValueDictionary[int, Any] | None = None


def callback(
    ret_type: Any,
    arg_types: list[Any],
    js_fn: Callable[..., Any],
    options: Any = None,
) -> Pointer:
    if not callable(js_fn):
        raise FFIError("callback requires a callable")

    adapter = get_global_adapter()
    normalized_ret = normalize_type(ret_type)
    normalized_args = [normalize_type(t) for t in arg_types]

    desc = adapter.register_callback(js_fn, normalized_ret, normalized_args)

    def _unregister():
        get_resource_registry().unregister_callback(desc)
        adapter.unregister_callback(desc)

    desc["_unregister"] = _unregister
    get_resource_registry().register_callback(desc)

    return Pointer(desc)
